### DISCUSSION TITLE
Better user experience for llamacpp-server

### DISCUSSION THREAD

**Discussion by PierreCarceller** [ggml-org/llama.cpp#10012] [Ideas]
Hello! 

I'm a llamacpp-server user. In particular the `openai` apis.

In a dream world I would like : 
1. Use a "message list" to interact with my LLM because this way of doing things is simple and human readable.
2. Be able to finely control generation with a regex, a choice list, a grammar, etc.

A bit like what you can do at [VLLM](https://docs.vllm.ai/en/v0.6.0/serving/openai_compatible_server.html#extra-parameters-for-chat-api)

I can think of 2 alternative solutions that are a little less practical but easier to set up.

Solution 1 : Create a `/message-format` endpoint that would apply the template chat to the list of messages sent and then use the `/completion` endpoint that already exists.

Solution 2 : Let the client format the message list on its own (with jinja for example) before using the `/completion` endpoint. But in this case, the server must provide access to the information needed to do the job on the client side (the BOS token, the EOS token...etc).  

I hope I haven't missed any important information.

**Comment by ggerganov** [comment]
1. We already support the `/chat/completions` endpoint that should be pretty much what you need. Have you tried it or is it missing some specific functionality?

2. We already support GBNF grammar for constrained generation.

Let me know if you need more help or if you have a specific example that you would like demonstrated.

**Reply by PierreCarceller** [reply]
First of all, thank you for your reply and more generally for the incredible work you do.

"We already support the `/chat/completions` endpoint" -> Yes, and that's what I'm currently using by passing JSON schemas.  
``` python
{
      "type": "json_object",
     "schema": Model.model_json_schema(),
}
```
But if I want to use a custom grammar, for example, I can't use the `/chat/completions` endpoint. Or at least I don't see it documented as supported. [here](https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#post-v1chatcompletions-openai-compatible-chat-completions-api)

**Reply by ExtReMLapin** [reply]
Not everything is documented, but I assure you it works, just give a try

**Reply by ggerganov** [reply]
@ExtReMLapin Can you show an example of using grammar with the chat endpoint? I haven't tried this combination yet, so not sure if it is working as expected. Will look into it soon though.

**Reply by ExtReMLapin** [reply]
@ggerganov 
[input_normal.json](https://github.com/user-attachments/files/17509963/input_normal.json)
[output_normal.json](https://github.com/user-attachments/files/17509964/output_normal.json)
[input_chat.json](https://github.com/user-attachments/files/17509965/input_chat.json)
[output_chat.json](https://github.com/user-attachments/files/17509966/output_chat.json)

**Reply by Soum-Soum** [reply]
Glad to read that! That's great news! Should update the ReadMe! I'll do my tests and propose a PR to update all this!
